test/functional/runner/index.js

```diff
@@ -0,0 +1,590 @@
+        }
+      });
+    });
+  });
+}
+
+// Test runner helpers
+function prepareDatabaseForSuite(suite, context) {
+  context.dbName = suite.database_name;
+  context.collectionName = suite.collection_name;
+
+  const db = context.sharedClient.db(context.dbName);
+  const coll = db.collection(context.collectionName);
+
+  return db
+    .admin()
+    .command({ killAllSessions: [] })
+    .catch(() => {}) // ignore any error from this
+    .then(() => coll.drop({ writeConcern: 'majority' }))
+    .catch(err => {
+      if (!err.message.match(/ns not found/)) throw err;
+    })
+    .then(() => {
+      if (suite.key_vault_data) {
+        const dataKeysCollection = context.sharedClient.db('admin').collection('datakeys');
+        return dataKeysCollection
+          .drop({ w: 'majority' })
+          .catch(err => {
+            if (!err.message.match(/ns not found/)) {
+              throw err;
+            }
+          })
+          .then(() => {
+            if (suite.key_vault_data.length) {
+              return dataKeysCollection.insertMany(suite.key_vault_data, { w: 'majority' });
+            }
+          });
+      }
+    })
+    .then(() => {
+      const options = { w: 'majority' };
+      if (suite.json_schema) {
+        options.validator = { $jsonSchema: suite.json_schema };
+      }
+
+      return db.createCollection(context.collectionName, options);
+    })
+    .then(() => {
+      if (suite.data && Array.isArray(suite.data) && suite.data.length > 0) {
+        return coll.insertMany(suite.data, { w: 'majority' });
+      }
+    })
+    .then(() => {
+      return context.runForAllClients(client => {
+        return client
+          .db(context.dbName)
+          .collection(context.collectionName)
+          .distinct('x')
+          .catch(() => {});
+      });
+    });
+}
+
+function parseSessionOptions(options) {
+  const result = Object.assign({}, options);
+  if (result.defaultTransactionOptions && result.defaultTransactionOptions.readPreference) {
+    result.defaultTransactionOptions.readPreference = normalizeReadPreference(
+      result.defaultTransactionOptions.readPreference.mode
+    );
+  }
+
+  return result;
+}
+
+const IGNORED_COMMANDS = new Set(['ismaster']);
+
+let displayCommands = false;
+function runTestSuiteTest(configuration, spec, context) {
+  const commandEvents = [];
+  const clientOptions = translateClientOptions(
+    Object.assign({ monitorCommands: true }, spec.clientOptions)
+  );
+
+  // test-specific client options
+  clientOptions.autoReconnect = false;
+  clientOptions.haInterval = 100;
+  clientOptions.useRecoveryToken = true;
+
+  // TODO: this should be configured by `newClient` and env variables
+  clientOptions.useUnifiedTopology = true;
+
+  const url = resolveConnectionString(configuration, spec);
+  const client = configuration.newClient(url, clientOptions);
+  return client.connect().then(client => {
+    context.testClient = client;
+    client.on('commandStarted', event => {
```

Comment:
do we track commandFailed or commandSucceeded events in any of our runners?

Reply:
We don't. It hasn't been a requirement yet, so it's not part of the unified runner. In the future, I imagine we have options to configure the test runner to enable features like that. KISS for now.